fixes bugs in handling of time varying R

stephaneghozzi opened the pull request:
reconhub/projections/issues/29
R_fix_within = TRUE now leads to a fixed R for each simulation and each time period

reconhub/projections/issues/28
- Cases y_t are now drawn from distributions with mean \sum_i R(i) * ws(t-i)  * y_i, with i < t and ws the PMF of the serial interval
- for better readability and to avoid possible mistakes: the matrix simulated counts out is now initialized only with initial incidence and results appended at each simulation step (instead of it being initiated with 0's for all simulations steps)

stephaneghozzi commented:
Fixes #28, fixes #29

thibautjombart commented:
Thanks for this! I get some errors on `devtools::check()` and also when running [this script](https://github.com/reconhub/trendbreaker/blob/master/inst/simulations.R). Can you check if you get the same?

Pasting the log of `devtools::check()` here:

<pre><font color="#CC0000">E  checking examples</font><font color="#06989A"> (478ms)</font>
   Running examples in ‘projections-Ex.R’ failed
   The error most likely occurred in:
   
   &gt; base::assign(&quot;.ptime&quot;, proc.time(), pos = &quot;CheckExEnv&quot;)
   &gt; ### Name: cumulate.projections
   &gt; ### Title: Compute cumulative projections
   &gt; ### Aliases: cumulate.projections
   &gt; 
   &gt; ### ** Examples
   &gt; 
   &gt; 
   &gt; if (require(distcrete) &amp;&amp;
   +     require(incidence)) {
   + 
   +   ## simulate basic epicurve
   +   dat &lt;- c(0, 2, 2, 3, 3, 5, 5, 5, 6, 6, 6, 6)
   +   i &lt;- incidence(dat)
   + 
   + 
   +   ## example with a function for SI
   +   si &lt;- distcrete(&quot;gamma&quot;, interval = 1L,
   +                   shape = 1.5,
   +                   scale = 2, w = 0)
   +   set.seed(1)
   +   pred_1 &lt;- project(i, runif(100, 0.8, 1.9), si, n_days = 30)
   +   plot_1 &lt;- plot(pred_1)
   + 
   +   ## cumulative predictions
   +   pred_1_cum &lt;- cumulate(pred_1)
   +   pred_1_cum
   +   plot(pred_1_cum)
   + }
   Loading required package: distcrete
   Loading required package: incidence
   Warning in rep(no, length.out = len) :
     &apos;x&apos; is NULL so the result will be NULL
   Error in ans[npos] &lt;- rep(no, length.out = len)[npos] : 
     replacement has length zero
   Calls: project -&gt; ifelse
   Execution halted
<font color="#4E9A06">✔</font>  <font color="#9E9E9E">checking for unstated dependencies in ‘tests’</font> ...
<font color="#9E9E9E">─  checking tests</font> ...
<font color="#CC0000">E</font><font color="#9E9E9E">  Running ‘testthat.R’</font><font color="#06989A"> (1.5s)</font>
   Running the tests in ‘tests/testthat.R’ failed.
   Last 13 lines of output:
     ══ testthat results  ═══════════════════════════════════════════════════════════
     [ OK: 27 | SKIPPED: 0 | WARNINGS: 26 | FAILED: 16 ]
     1. Error: Accessors return the right thing (@test-accessors.R#22) 
     2. Error: Test round trip (@test-build_projections.R#23) 
     3. Error: Test dates default (@test-build_projections.R#48) 
     4. Error: Test against reference results (@test-conversion.R#23) 
     5. Error: Printing as planned (@test-cumulate.R#23) 
     6. Error: Merging works (@test-merge_add_projections.R#16) 
     7. Error: Merging works (@test-merge_projections.R#15) 
     8. Error: Test against reference results (@test-plots.R#23) 
     9. Error: Printing as planned (@test-print.R#23) 
     1. ...
     
     Error: testthat unit tests failed
     Execution halted
<font color="#4E9A06">✔</font>  <font color="#9E9E9E">checking for detritus in the temp directory</font> ...
   
   See
     ‘/tmp/Rtmpqr6i8O/projections.Rcheck/00check.log’
   for details.
   
<font color="#06989A">── R CMD check results ────────────────────────────────── projections 0.4.1 ────</font>
Duration: 51.4s

<font color="#CC0000">❯ checking examples ... ERROR</font>
  Running examples in ‘projections-Ex.R’ failed
  The error most likely occurred in:
  
  &gt; base::assign(&quot;.ptime&quot;, proc.time(), pos = &quot;CheckExEnv&quot;)
  &gt; ### Name: cumulate.projections
  &gt; ### Title: Compute cumulative projections
  &gt; ### Aliases: cumulate.projections
  &gt; 
  &gt; ### ** Examples
  &gt; 
  &gt; 
  &gt; if (require(distcrete) &amp;&amp;
  +     require(incidence)) {
  + 
  +   ## simulate basic epicurve
  +   dat &lt;- c(0, 2, 2, 3, 3, 5, 5, 5, 6, 6, 6, 6)
  +   i &lt;- incidence(dat)
  + 
  + 
  +   ## example with a function for SI
  +   si &lt;- distcrete(&quot;gamma&quot;, interval = 1L,
  +                   shape = 1.5,
  +                   scale = 2, w = 0)
  +   set.seed(1)
  +   pred_1 &lt;- project(i, runif(100, 0.8, 1.9), si, n_days = 30)
  +   plot_1 &lt;- plot(pred_1)
  + 
  +   ## cumulative predictions
  +   pred_1_cum &lt;- cumulate(pred_1)
  +   pred_1_cum
  +   plot(pred_1_cum)
  + }
  Loading required package: distcrete
  Loading required package: incidence
  Warning in rep(no, length.out = len) :
    &apos;x&apos; is NULL so the result will be NULL
  Error in ans[npos] &lt;- rep(no, length.out = len)[npos] : 
    replacement has length zero
  Calls: project -&gt; ifelse
  Execution halted

<font color="#CC0000">❯ checking tests ...</font>
  See below...

<font color="#06989A">── Test failures ───────────────────────────────────────────────── testthat ────</font>

&gt; library(testthat)
&gt; library(vdiffr)
&gt; library(distcrete)
&gt; library(incidence)
&gt; library(projections)
&gt; library(outbreaks)
&gt; library(magrittr)

Attaching package: &apos;magrittr&apos;

The following objects are masked from &apos;package:testthat&apos;:

    equals, is_less_than, not

&gt; 
&gt; test_check(&quot;projections&quot;)
── 1. Error: Accessors return the right thing (@test-accessors.R#22)  ──────────
replacement has length zero
Backtrace:
 1. projections::project(i, runif(100, 0.8, 1.9), si, n_days = 30)
 2. base::ifelse(is.finite(time_change), time_change, NULL)

── 2. Error: Test round trip (@test-build_projections.R#23)  ───────────────────
replacement has length zero
Backtrace:
 1. projections::project(i, runif(100, 0.8, 1.9), si, n_days = 30)
 2. base::ifelse(is.finite(time_change), time_change, NULL)

── 3. Error: Test dates default (@test-build_projections.R#48)  ────────────────
replacement has length zero
Backtrace:
 1. projections::project(i, runif(100, 0.8, 1.9), si, n_days = 30)
 2. base::ifelse(is.finite(time_change), time_change, NULL)

── 4. Error: Test against reference results (@test-conversion.R#23)  ───────────
replacement has length zero
Backtrace:
 1. projections::project(i, runif(100, 0.8, 1.9), si, n_days = 30)
 2. base::ifelse(is.finite(time_change), time_change, NULL)

── 5. Error: Printing as planned (@test-cumulate.R#23)  ────────────────────────
replacement has length zero
Backtrace:
 1. projections::project(i, runif(100, 0.8, 1.9), si, n_days = 30)
 2. base::ifelse(is.finite(time_change), time_change, NULL)

── 6. Error: Merging works (@test-merge_add_projections.R#16)  ─────────────────
replacement has length zero
Backtrace:
 1. projections::project(...)
 2. base::ifelse(is.finite(time_change), time_change, NULL)

── 7. Error: Merging works (@test-merge_projections.R#15)  ─────────────────────
replacement has length zero
Backtrace:
 1. projections::project(...)
 2. base::ifelse(is.finite(time_change), time_change, NULL)

── 8. Error: Test against reference results (@test-plots.R#23)  ────────────────
replacement has length zero
Backtrace:
 1. projections::project(i, runif(100, 0.8, 1.9), si, n_days = 30)
 2. base::ifelse(is.finite(time_change), time_change, NULL)

── 9. Error: Printing as planned (@test-print.R#23)  ───────────────────────────
replacement has length zero
Backtrace:
 1. projections::project(i, runif(100, 0.8, 1.9), si, n_days = 30)
 2. base::ifelse(is.finite(time_change), time_change, NULL)

── 10. Error: Projections can be performed for a single day (@test-project.R#15)
replacement has length zero
Backtrace:
 1. projections::project(...)
 2. base::ifelse(is.finite(time_change), time_change, NULL)

── 11. Error: Projections can be performed for a single day (@test-project.R#32)
replacement has length zero
Backtrace:
 1. projections::project(...)
 2. base::ifelse(is.finite(time_change), time_change, NULL)

── 12. Error: Projections can be performed for a single day and single simulatio
replacement has length zero
Backtrace:
 1. projections::project(...)
 2. base::ifelse(is.finite(time_change), time_change, NULL)

── 13. Error: Test against reference results (@test-project.R#77)  ─────────────
replacement has length zero
Backtrace:
 1. projections::project(i, runif(100, 0.8, 1.9), si, n_days = 30)
 2. base::ifelse(is.finite(time_change), time_change, NULL)

── 14. Error: Test that dates start when needed (@test-project.R#140)  ─────────
replacement has length zero
Backtrace:
 1. projections::project(i, runif(100, 0.8, 1.9), si, n_days = 30)
 2. base::ifelse(is.finite(time_change), time_change, NULL)

── 15. Error: Test against reference results - numeric dates (@test-subset.R#23)
replacement has length zero
Backtrace:
 1. projections::project(i, runif(100, 0.8, 1.9), si, n_days = 30)
 2. base::ifelse(is.finite(time_change), time_change, NULL)

── 16. Error: Test against reference results - Date dates (@test-subset.R#59)  ─
replacement has length zero
Backtrace:
 1. projections::project(i, runif(100, 0.8, 1.9), si, n_days = 30)
 2. base::ifelse(is.finite(time_change), time_change, NULL)

══ testthat results  ═══════════════════════════════════════════════════════════
[ OK: 27 | SKIPPED: 0 | WARNINGS: 26 | FAILED: 16 ]
1. Error: Accessors return the right thing (@test-accessors.R#22) 
2. Error: Test round trip (@test-build_projections.R#23) 
3. Error: Test dates default (@test-build_projections.R#48) 
4. Error: Test against reference results (@test-conversion.R#23) 
5. Error: Printing as planned (@test-cumulate.R#23) 
6. Error: Merging works (@test-merge_add_projections.R#16) 
7. Error: Merging works (@test-merge_projections.R#15) 
8. Error: Test against reference results (@test-plots.R#23) 
9. Error: Printing as planned (@test-print.R#23) 
1. ...

Error: testthat unit tests failed
Execution halted

<font color="#CC0000">2 errors ✖</font> | <font color="#4E9A06">0 warnings ✔</font> | <font color="#4E9A06">0 notes ✔</font>
&gt; 
</pre>

stephaneghozzi commented:
The last commits corrects a mistake I had introduced while summing up with serial interval pmf 'si': the first term of si corresponds to an interval of 0, which doesn't play a role here.

Moreover, the text of my previous commit was correct but somewhat misleading: in the code 'ws' is the reverse of the pmf of the serial interval 'si' overall the whole simulation time 0:t_max, so the formula would best be written as:
\sum_i R(i) * si(t-i) * y_i = \sum_i R(i) * ws((t_max-i):(t_max-1)) * y_i

stephaneghozzi commented:
> Thanks for this! I get some errors on `devtools::check()` and also when running [this script](https://github.com/reconhub/trendbreaker/blob/master/inst/simulations.R). Can you check if you get the same?

I'll have a look.

stephaneghozzi commented:
Thanks for pointing out the errors. Indeed I have been mystified by `ifelse()` once again (problem with different lengths of output, I should learn), and replaced it with a loop.

So now [simulations.R](https://github.com/reconhub/trendbreaker/blob/master/inst/simulations.R) should be running... if two lines of it are commented/corrected: 
l141: `library(cadet) # presumably caret ;)`
l246 (the last): `list_data <- lapply(list_res, add_reporting, f_reporting) # list_res is not defined, presumably list_res_free_k or list_res_fixed_k`

I'll look at `check()` later, probably tomorrow...

thibautjombart commented:
Great, thanks! I confirm the simulations in *trendbreaker* are now working as expected. 

I still get the following errors with `devtools::check()`:
```r
E  checking examples (1.5s)
   Running examples in ‘projections-Ex.R’ failed
   The error most likely occurred in:
   
   > base::assign(".ptime", proc.time(), pos = "CheckExEnv")
   > ### Name: merge_projections
   > ### Title: Merge a list of projections objects
   > ### Aliases: merge_projections
   > 
   > ### ** Examples
   > 
   > 
   > ## generate toy data
   > dates <- Sys.Date() + c(0, 0, 2, 5, 6, 6, 7)
   > i <- incidence::incidence(dates)
   > si <- c(0.2, 0.5, 0.2, 0.1)
   > R0 <- 3.5
   > 
   > ## make several projections objects
   > x <- lapply(1:10,
   +             function(j)
   +               project(x = i,
   +                       si = si,
   +                       R = R0,
   +                       n_sim = 2 * j,
   +                       R_fix_within = TRUE,
   +                       n_days = j,
   +                       model = "poisson"
   +                       ))
   Error in rbind(R_t, current_R_t) : 
     number of columns of matrices must match (see arg 2)
   Calls: lapply -> FUN -> project -> rbind
   Execution halted
✔  checking for unstated dependencies in ‘tests’
─  checking tests
E  Running ‘testthat.R’ (8.5s)
   Running the tests in ‘tests/testthat.R’ failed.
   Last 13 lines of output:
     ══ testthat results  ═══════════════════════════════════════════════════════════
     [ OK: 45 | SKIPPED: 0 | WARNINGS: 11 | FAILED: 29 ]
     1. Failure: Test against reference results (@test-conversion.R#27) 
     2. Failure: Test against reference results (@test-conversion.R#32) 
     3. Error: Merging works (@test-merge_add_projections.R#16) 
     4. Error: Merging works (@test-merge_projections.R#15) 
     5. Failure: Test against reference results (@test-plots.R#25) 
     6. Failure: Test against reference results (@test-plots.R#41) 
     7. Failure: Test against reference results (@test-plots.R#42) 
     8. Failure: Test against reference results (@test-plots.R#44) 
     9. Failure: Test against reference results (@test-plots.R#46) 
     1. ...
     
     Error: testthat unit tests failed
     Execution halted
✔  checking for detritus in the temp directory
   
   See
     ‘/tmp/RtmpdQ3iTT/projections.Rcheck/00check.log’
   for details.
   
── R CMD check results ────────────────────────────────── projections 0.4.1 ────
Duration: 31.8s

❯ checking examples ... ERROR
  Running examples in ‘projections-Ex.R’ failed
  The error most likely occurred in:
  
  > base::assign(".ptime", proc.time(), pos = "CheckExEnv")
  > ### Name: merge_projections
  > ### Title: Merge a list of projections objects
  > ### Aliases: merge_projections
  > 
  > ### ** Examples
  > 
  > 
  > ## generate toy data
  > dates <- Sys.Date() + c(0, 0, 2, 5, 6, 6, 7)
  > i <- incidence::incidence(dates)
  > si <- c(0.2, 0.5, 0.2, 0.1)
  > R0 <- 3.5
  > 
  > ## make several projections objects
  > x <- lapply(1:10,
  +             function(j)
  +               project(x = i,
  +                       si = si,
  +                       R = R0,
  +                       n_sim = 2 * j,
  +                       R_fix_within = TRUE,
  +                       n_days = j,
  +                       model = "poisson"
  +                       ))
  Error in rbind(R_t, current_R_t) : 
    number of columns of matrices must match (see arg 2)
  Calls: lapply -> FUN -> project -> rbind
  Execution halted

❯ checking tests ...
  See below...

── Test failures ───────────────────────────────────────────────── testthat ────

> library(testthat)
> library(vdiffr)
> library(distcrete)
> library(incidence)
> library(projections)
> library(outbreaks)
> library(magrittr)

Attaching package: 'magrittr'

The following objects are masked from 'package:testthat':

    equals, is_less_than, not

> 
> test_check("projections")
── 1. Failure: Test against reference results (@test-conversion.R#27)  ─────────
`df_1` has changed from known value recorded in 'rds/df_1.rds'.
Component "sim_1": Mean relative difference: 0.915493
Component "sim_2": Mean relative difference: 0.7777778
Component "sim_3": Mean relative difference: 4.464286
Component "sim_4": Mean relative difference: 1.195652
Component "sim_5": Mean relative difference: 1.118644
Component "sim_6": Mean relative difference: 0.4123711
Component "sim_7": Mean relative difference: 0.7820513
Component "sim_8": Mean relative difference: 0.4903226
Component "sim_9": Mean relative difference: 2.142857
...

── 2. Failure: Test against reference results (@test-conversion.R#32)  ─────────
`df_2` has changed from known value recorded in 'rds/df_2.rds'.
Component "incidence": Mean relative difference: 0.9841556

── 3. Error: Merging works (@test-merge_add_projections.R#16)  ─────────────────
number of columns of matrices must match (see arg 2)
Backtrace:
 1. projections::project(...)
 2. base::rbind(R_t, current_R_t)

── 4. Error: Merging works (@test-merge_projections.R#15)  ─────────────────────
number of columns of matrices must match (see arg 2)
Backtrace:
 1. projections::project(...)
 2. base::rbind(R_t, current_R_t)

── 5. Failure: Test against reference results (@test-plots.R#25)  ──────────────
Figures don't match: basic-example-plot.svg


── 6. Failure: Test against reference results (@test-plots.R#41)  ──────────────
Figures don't match: evd-proj.svg


── 7. Failure: Test against reference results (@test-plots.R#42)  ──────────────
Figures don't match: evd-proj-box-no-outliers.svg


── 8. Failure: Test against reference results (@test-plots.R#44)  ──────────────
Figures don't match: evd-proj-no-ribbon.svg


── 9. Failure: Test against reference results (@test-plots.R#46)  ──────────────
Figures don't match: evd-proj-no-box-custom-lines.svg


── 10. Failure: Test against reference results (@test-plots.R#49)  ─────────────
Figures don't match: evd-proj-red-box.svg


── 11. Failure: Test against reference results (@test-plots.R#51)  ─────────────
Figures don't match: evd-proj-box-only.svg


── 12. Failure: Test against reference results (@test-plots.R#54)  ─────────────
Figures don't match: evd-proj-ribbon-only.svg


── 13. Failure: Test against reference results (@test-plots.R#56)  ─────────────
Figures don't match: evd-proj-red-ribbon.svg


── 14. Failure: Test against reference results (@test-plots.R#59)  ─────────────
Figures don't match: evd-proj-full-red-ribbon-narrow-range.svg


── 15. Failure: Test against reference results (@test-plots.R#67)  ─────────────
Figures don't match: evd-proj-with-incidence-incidence.svg


── 16. Failure: Test against reference results (@test-plots.R#69)  ─────────────
Figures don't match: evd-proj-with-incidence-incidence-no-box.svg


── 17. Failure: Test against reference results (@test-plots.R#72)  ─────────────
Figures don't match: evd-proj-with-incidence-incidence-box-only.svg


── 18. Failure: Test against reference results (@test-plots.R#80)  ─────────────
Figures don't match: evd-proj-with-incidence-incidence-and-custom.svg


── 19. Error: Projections can be performed for a single day (@test-project.R#15)
number of columns of matrices must match (see arg 2)
Backtrace:
 1. projections::project(...)
 2. base::rbind(R_t, current_R_t)

── 20. Failure: Test against reference results (@test-project.R#79)  ───────────
`pred_1` has changed from known value recorded in 'rds/pred_1.rds'.
Mean relative difference: 0.9841556

── 21. Failure: Test against reference results (@test-project.R#91)  ───────────
`pred_2` has changed from known value recorded in 'rds/pred_2.rds'.
Mean relative difference: 1.062904

── 22. Failure: Test against reference results (@test-project.R#104)  ──────────
`pred_3` has changed from known value recorded in 'rds/pred_3.rds'.
Mean relative difference: 0.7419163

── 23. Failure: Test against reference results (@test-project.R#118)  ──────────
`pred_4` has changed from known value recorded in 'rds/pred_4.rds'.
Mean relative difference: 0.7829049

── 24. Failure: Test against reference results - numeric dates (@test-subset.R#3
`subset_1` has changed from known value recorded in 'rds/subset_1.rds'.
Mean relative difference: 0.9571429

── 25. Failure: Test against reference results - numeric dates (@test-subset.R#3
`subset_2` has changed from known value recorded in 'rds/subset_2.rds'.
Mean relative difference: 0.9898214

── 26. Failure: Test against reference results - numeric dates (@test-subset.R#3
`subset_3` has changed from known value recorded in 'rds/subset_3.rds'.
Mean relative difference: 0.9134002

── 27. Failure: Test against reference results - Date dates (@test-subset.R#69) 
`subset_1` has changed from known value recorded in 'rds/subset_Date_1.rds'.
Mean relative difference: 0.9571429

── 28. Failure: Test against reference results - Date dates (@test-subset.R#70) 
`subset_2` has changed from known value recorded in 'rds/subset_Date_2.rds'.
Mean relative difference: 0.9898214

── 29. Failure: Test against reference results - Date dates (@test-subset.R#71) 
`subset_3` has changed from known value recorded in 'rds/subset_Date_3.rds'.
Mean relative difference: 0.9134002

══ testthat results  ═══════════════════════════════════════════════════════════
[ OK: 45 | SKIPPED: 0 | WARNINGS: 11 | FAILED: 29 ]
1. Failure: Test against reference results (@test-conversion.R#27) 
2. Failure: Test against reference results (@test-conversion.R#32) 
3. Error: Merging works (@test-merge_add_projections.R#16) 
4. Error: Merging works (@test-merge_projections.R#15) 
5. Failure: Test against reference results (@test-plots.R#25) 
6. Failure: Test against reference results (@test-plots.R#41) 
7. Failure: Test against reference results (@test-plots.R#42) 
8. Failure: Test against reference results (@test-plots.R#44) 
9. Failure: Test against reference results (@test-plots.R#46) 
1. ...

Error: testthat unit tests failed
Execution halted

2 errors ✖ | 0 warnings ✔ | 0 notes ✔
> 
```

stephaneghozzi commented:
I had one more bug I could fix. Now it is as you said: errors only come from comparison with expected results or plots.

I'll look closer into the results and plots and then we can discuss whether the tests should be rewritten...

But already an update, with the following tests removed for both versions: [test-conversion.R](https://github.com/reconhub/projections/blob/master/tests/testthat/test-conversion.R), [test-plots.R](https://github.com/reconhub/projections/blob/master/tests/testthat/test-plots.R), [test-subset.R](https://github.com/reconhub/projections/blob/master/tests/testthat/test-subset.R), and the series of "Test against reference results" in [test-project.R](https://github.com/reconhub/projections/blob/master/tests/testthat/test-project.R)

**reconhub/projections:master**

devtools::check()
```
...
-- R CMD check results ------------------------------------- projections 0.4.1 ----
Duration: 28.2s

0 errors √ | 0 warnings √ | 0 notes √
```
devtools::test()
```
Loading projections
Testing projections
√ |  OK F W S | Context
√ |   2   1   | Test accessors
--------------------------------------------------------------------------------
test-accessors.R:3: warning: (unknown)
non-uniform 'Rounding' sampler used
--------------------------------------------------------------------------------
√ |   7   1   | Test build_projections
--------------------------------------------------------------------------------
test-build_projections.R:3: warning: (unknown)
non-uniform 'Rounding' sampler used
--------------------------------------------------------------------------------
√ |   2   1   | Test printing
--------------------------------------------------------------------------------
test-cumulate.R:3: warning: (unknown)
non-uniform 'Rounding' sampler used
--------------------------------------------------------------------------------
√ |   6       | Test internals
√ |  12   1   | Test merge_add_projections [0.2 s]
--------------------------------------------------------------------------------
test-merge_add_projections.R:3: warning: (unknown)
non-uniform 'Rounding' sampler used
--------------------------------------------------------------------------------
√ |  10   1   | Test merge_projections
--------------------------------------------------------------------------------
test-merge_projections.R:3: warning: (unknown)
non-uniform 'Rounding' sampler used
--------------------------------------------------------------------------------
√ |   4   1   | Test printing
--------------------------------------------------------------------------------
test-print.R:3: warning: (unknown)
non-uniform 'Rounding' sampler used
--------------------------------------------------------------------------------
√ |  15   1   | Test project function
--------------------------------------------------------------------------------
test-project.R:3: warning: (unknown)
non-uniform 'Rounding' sampler used
--------------------------------------------------------------------------------

== Results =====================================================================
Duration: 1.0 s

OK:       58
Failed:   0
Warnings: 7
Skipped:  0
```

**stephaneghozzi/projections:bugfix-project-R_t**

devtools::check()
```
...
-- R CMD check results ------------------------------------- projections 0.4.1 ----
Duration: 27.7s

0 errors √ | 0 warnings √ | 0 notes √
```
devtools::test()
```
Loading projections
Testing projections
√ |  OK F W S | Context
√ |   2   1   | Test accessors
--------------------------------------------------------------------------------
test-accessors.R:3: warning: (unknown)
non-uniform 'Rounding' sampler used
--------------------------------------------------------------------------------
√ |   7   1   | Test build_projections
--------------------------------------------------------------------------------
test-build_projections.R:3: warning: (unknown)
non-uniform 'Rounding' sampler used
--------------------------------------------------------------------------------
√ |   2   1   | Test printing
--------------------------------------------------------------------------------
test-cumulate.R:3: warning: (unknown)
non-uniform 'Rounding' sampler used
--------------------------------------------------------------------------------
√ |   6       | Test internals
√ |  12   1   | Test merge_add_projections [0.2 s]
--------------------------------------------------------------------------------
test-merge_add_projections.R:3: warning: (unknown)
non-uniform 'Rounding' sampler used
--------------------------------------------------------------------------------
√ |  10   1   | Test merge_projections
--------------------------------------------------------------------------------
test-merge_projections.R:3: warning: (unknown)
non-uniform 'Rounding' sampler used
--------------------------------------------------------------------------------
√ |   4   1   | Test printing
--------------------------------------------------------------------------------
test-print.R:3: warning: (unknown)
non-uniform 'Rounding' sampler used
--------------------------------------------------------------------------------
√ |  15   1   | Test project function
--------------------------------------------------------------------------------
test-project.R:3: warning: (unknown)
non-uniform 'Rounding' sampler used
--------------------------------------------------------------------------------

== Results =====================================================================
Duration: 1.1 s

OK:       58
Failed:   0
Warnings: 7
Skipped:  0
```

thibautjombart commented:
Thanks for the additional fixes! I have started looking at differences with reference objects triggering the errors on `devtools::test()`. First thing is, it was probably a bad idea in the first place to test against reference objects. We'll probably do away with these gradually as things progress.

This said, they may reflect actual issues. The first one is that something has changed in `pred_1` in the following code:

```r
## simulate basic epicurve
dat <- as.Date("2001-01-01") + c(0, 2, 2, 3, 3, 5, 5, 5, 6, 6, 6, 6)
i <- incidence::incidence(dat)

## example with a function for SI
si <- distcrete::distcrete("gamma", interval = 1L,
                               shape = 1.5,
                               scale = 2, w = 0)
    
set.seed(1)
pred_1 <- project(i, runif(100, 0.8, 1.9), si, n_days = 30)
```

Seeing that the code does not use time-varrying R, it seems odd that the patch alters this output. Any idea? I suspect once we get to the bottom of that one the rest will be solved quickly.

thibautjombart commented on R/project.R:
```diff
@@ -253,59 +254,75 @@ project <- function(x, R, si, n_sim = 100, n_days = 7,
+      (R_t[seq_len(i-1), , drop = FALSE] * out)
     ## lambda <- lambda / reporting
     if (model == "poisson") {
-      out[i, ] <- stats::rpois(n_sim, current_R * lambda)
```
This code is confusing (and also my fault). It uses data up to time `i` to compute the force of infection and new cases at time `i`. I think it works because `tail(ws, 1)` should normally be 0 (this is the pmf for 0 days, i.e. infectiousness on the day of onset), and the corresponding nb of cases is 0 due to how `out` is initialised. Still, it would be better to use data up to time `i` to define `infectiousness on time `i+1`.

stephaneghozzi replied:
I agree that computing `i+1`does sound more intuitive...
I don't know enough of epidemiology to say whether `tail(ws, 1)` resp. `si[1]` should be 0, but for the reason you mention, it doesn't play a role here...

thibautjombart commented on R/project.R:
```diff
@@ -253,59 +254,75 @@ project <- function(x, R, si, n_sim = 100, n_days = 7,
-    }
-    lambda <- utils::tail(ws, i) %*% out[seq_len(i), , drop = FALSE]
+
+    lambda_R <- ws[(length(ws)-(i-1)):(length(ws)-1)] %*%
```
I think the reversing here may be wrong, as `ws` is already the reverse of the pmf of the serial interval.

stephaneghozzi replied:
I think this is correct in both versions, although maybe writing something like `si[(i-1):1]` and not defining a `ws` might be better understandable, but surely it's a matter of taste... `ws` is not reversed, just taking the last elements (the last one excluded in the new version).

thibautjombart commented:
Here is a proposed gist for the force of infection:
https://gist.github.com/thibautjombart/ba51f3bf5a2104e21a09550df24ba7fc

Let me know what you think. If you are happy with this kind of approach, I suggest we add this as an internal (non-exported) function and use it in `project`. One advantage is that this function can be tested separately.

stephaneghozzi commented:
> I fear both calculations might be wrong (sic).
I think both are correct wrt pmf 🙂

> Note that ws is the reversed pmf of the serial interval si so that e.g. tail(ws, 3) is si[3:1], which is effectively [si(0) si(1) si(2)] (the first element of si is the mass for day 0).
yes, I first didn't pay attention but corrected it in stephaneghozzi/projections/commit/f964dca828ab9d626542451aa2a4c324cf4b44fa
For the reasons you mention above I preferred appending values to out rather than initializing with 0's... then don't have to to worry about initializing, it makes clearer which values of `ws` are used, and you don't worry about which row you're computing, whether `i` or `i+1`... although the latter is still better when `i` is used 

> Let me know what you think.
I will 🙂

stephaneghozzi commented:
> Seeing that the code does not use time-varrying R, it seems odd that the patch alters this output. Any idea? 

I think it's just the randomly drawing from the Poisson distribution, the seed is not set there...? I'll upload the comparison plots later today, they look very similar... But I'll look again at the handling of the seed...

stephaneghozzi commented:
> I think it's just the randomly drawing from the Poisson distribution, the seed is not set there...? I'll upload the comparison plots later today, they look very similar... But I'll look again at the handling of the seed...

Upon reflection, of course the seed applies to all random numbers,  otherwise the test would never work...

stephaneghozzi commented:
I added the function `compute_force_infection()` you wrote, added a description for the argument `t` and use it in `project()` (which was modified accordingly).

stephaneghozzi commented:
Now for some plots...
Regarding issue #28 
![image](https://user-images.githubusercontent.com/51886077/85206702-12926500-b324-11ea-8956-db7fa28dba85.png)
and issue #29 
![image](https://user-images.githubusercontent.com/51886077/85206710-1e7e2700-b324-11ea-97f4-1b6f01c7af64.png)

stephaneghozzi commented:
About the tests that didn't work, I generated my own references and plotted them to compare with the original ones. I added the code to do so to the test files.

Script to read the different references and plot them or convert them to png:
```r
library(projections)
library(ggplot2)
library(patchwork)
library(tidyr)
library(dplyr)
library(gridExtra)
library(rsvg)

# Comparison of test references for package projections 
# for master vs. bugfix-project-R_t
#
# ----------------------------------------------------------------------------
#
# .rds files produced by e.g.
# dat <- c(0, 2, 2, 3, 3, 5, 5, 5, 6, 6, 6, 6)
# i <- incidence::incidence(dat)
# si <- distcrete::distcrete("gamma", interval = 1L,
#   shape = 1.5,
#   scale = 2, w = 0)
# set.seed(1)
# pred_1 <- project(i, runif(100, 0.8, 1.9), si, n_days = 30)
# saveRDS(pred_1, file='rds/pred_1.rds')
#
# pred_* are computed in test-plot.R
#   N.B. for pred_*, serial interval is     
#     si <- distcrete::distcrete("gamma", interval = 1L, shape = 1.5, 
#       scale = 2, w = 0)
#   which has si$d(0)=0.20 and mean(si$d(0:100)) = 0.01
# df_* are computed in test-conversion.R
# subset_* are computed in test-subset.R
# 
# ----------------------------------------------------------------------------
# .svg files produced in test-plots.R by e.g.
# library(ggplot2)
# library(svglite)
# pred_1 <- project(i, runif(100, 0.8, 1.9), si, n_days = 30)
# plot_1 <- plot(pred_1)
# ggsave('../figs/test-project-function/basic-example-plot.svg',
#   plot=plot_1)

# plot rds files

path_rds_master <- '../projections-master/tests/testthat/rds/'
path_rds_Rt <- '../projections-Rt/tests/testthat/rds/'
for (frds in c('pred_', 'df_', 'subset_', 'subset_Date_')) {
  plot_list <- list()
  j <- 1
  for (i in 1:3) {
    if (!(i==3 & frds=='df_')) {
      rds_master <- readRDS(paste0(path_rds_master, frds, i, '.rds'))
      rds_Rt <- readRDS(paste0(path_rds_Rt, frds, i, '.rds'))
      date_vars <- c('dates','date')
      if (any(date_vars %in% names(rds_master))) {
        dvar <- date_vars[date_vars %in% names(rds_master)]
        idates <- which(names(rds_master)==dvar)
        ymax <- max(rds_master[,-idates], rds_Rt[,-idates])
        if (i==1) {
          rds_master1 <- rds_master %>% 
            pivot_longer(cols=-dvar, names_to='sim', values_to='incidence')
          rds_master1$sim <- as.factor(rds_master1$sim)
          rds_Rt1 <- rds_Rt %>% 
            pivot_longer(cols=-dvar, names_to='sim', values_to='incidence')
          rds_Rt1$sim <- as.factor(rds_Rt1$sim)
        } else {
          rds_master1 <- rds_master
          rds_master1$sim <- as.factor(rds_master1$sim)
          rds_Rt1 <- rds_Rt
          rds_Rt1$sim <- as.factor(rds_Rt1$sim)
        }
        p_master <- 
          ggplot(rds_master1, aes_string(x=dvar, y='incidence', color='sim')) + 
          geom_point(alpha=0.3) + 
          ylim(c(0,ymax)) + 
          ggtitle(paste0(frds, i, '.rds (original)')) + 
          theme(legend.position = 'none')
        p_Rt <- 
          ggplot(rds_Rt1, aes_string(x=dvar, y='incidence', color='sim')) + 
          geom_point(alpha=0.3) + 
          ylim(c(0,ymax)) + 
          ggtitle(paste0(frds, i, '.rds (new)')) + 
          theme(legend.position = 'none')
      } else {
        ymax <- max(rds_master, rds_Rt)
        p_master <- plot(rds_master) + ylim(c(0,ymax)) + 
          ggtitle(paste0(frds, i, '.rds (original)'))
        p_Rt <- plot(rds_Rt) + ylim(c(0,ymax)) + 
          ggtitle(paste0(frds, i, '.rds (new)'))
      }
      plot_list[[j]] <- p_master
      plot_list[[j+1]] <- p_Rt
      j <- j+2
    }
  }
  plot_rds_combined <- arrangeGrob(grobs=plot_list, ncol=2)
  ggsave(paste0('rds-plots/plot_', frds, 'rds_combined.png'), plot=plot_rds_combined, 
    width=30, height=i*10, units='cm', dpi=200)
}

# convert svg files to png

path_svg <- list(
  master = '../projections-master/tests/figs/test-project-function/',
  Rt = '../projections-Rt/tests/figs/test-project-function/'
)
svg_file_names <- c('basic-example-plot', 'evd-proj-box-no-outliers',
  'evd-proj-box-only', 'evd-proj-full-red-ribbon-narrow-range',
  'evd-proj-no-box-custom-lines', 'evd-proj-no-ribbon', 'evd-proj-red-box',
  'evd-proj-red-ribbon', 'evd-proj-ribbon-only',
  'evd-proj-with-incidence-incidence-and-custom',
  'evd-proj-with-incidence-incidence-box-only',
  'evd-proj-with-incidence-incidence-no-box','evd-proj-with-incidence-incidence',
  'evd-proj')

for (fsvg in svg_file_names) {
  for (branch in c('master', 'Rt'))
  rsvg_png(paste0(path_svg[[branch]], fsvg, '.svg'), 
    paste0('svg2png/', fsvg, '-', branch, '.png'), width=1000)
}
```

stephaneghozzi commented:
For the references for which tests failed and that are rds files (from test-conversion.R, test-subset.R, and the series of "Test against reference results" in test-project.R):
![image](https://user-images.githubusercontent.com/51886077/85206935-c5af8e00-b325-11ea-9987-b3a0f3b040b4.png)
![image](https://user-images.githubusercontent.com/51886077/85206938-c9dbab80-b325-11ea-8e25-d8091cfcdc1d.png)
![image](https://user-images.githubusercontent.com/51886077/85206940-ce07c900-b325-11ea-8753-6b92d8396989.png)
![image](https://user-images.githubusercontent.com/51886077/85206945-d3fdaa00-b325-11ea-9dd6-feef938339fa.png)

I am not sure what's going on with `df_*` as there are many results for each date and simulation (color = simulation number).

I still don't know why `pred_1` is not exactly the same between the two versions.

The abrupt change for `pred_3` is due to the fact that the serial interval selected has mean 0.01 (see comment in code in the previous comment).

stephaneghozzi commented:
For the references that are plots (svg files):

1. even when I generated them myself (with `ggsave("filename.svg")`), I get an error from `vdiffr::expect_doppelganger`... I don't know why. That'd be great if you'd have an advice and/or could look how it works on your side.

2. Otherwise the plots are very similar (white background = original, grey background = new):
![image](https://user-images.githubusercontent.com/51886077/85207371-af570180-b328-11ea-929d-b2aaa1d51dfd.png)
![image](https://user-images.githubusercontent.com/51886077/85207373-b251f200-b328-11ea-8da3-c0aef96d1679.png)
![image](https://user-images.githubusercontent.com/51886077/85207375-b67e0f80-b328-11ea-8cef-a1271b11ae25.png)
![image](https://user-images.githubusercontent.com/51886077/85207382-b8e06980-b328-11ea-9c4c-34261cc2d8c6.png)
![image](https://user-images.githubusercontent.com/51886077/85207386-bc73f080-b328-11ea-8175-a71f0fadd62b.png)
![image](https://user-images.githubusercontent.com/51886077/85207389-c0077780-b328-11ea-844f-9f6e2102e6ee.png)
![image](https://user-images.githubusercontent.com/51886077/85207391-c269d180-b328-11ea-8503-d4829d1013da.png)
![image](https://user-images.githubusercontent.com/51886077/85207394-c5fd5880-b328-11ea-8671-8fdd79656bb6.png)
![image](https://user-images.githubusercontent.com/51886077/85207395-ca297600-b328-11ea-9fcf-172bb47548bc.png)
![image](https://user-images.githubusercontent.com/51886077/85207399-cd246680-b328-11ea-8e6b-4bab3338fdd1.png)
![image](https://user-images.githubusercontent.com/51886077/85207403-d0b7ed80-b328-11ea-9454-4910b3fc840a.png)
![image](https://user-images.githubusercontent.com/51886077/85207407-d31a4780-b328-11ea-98e7-fdb10732c5ff.png)
![image](https://user-images.githubusercontent.com/51886077/85207411-d57ca180-b328-11ea-99d4-57953e920d98.png)
![image](https://user-images.githubusercontent.com/51886077/85207413-d8779200-b328-11ea-919c-9a475da096f6.png)
![image](https://user-images.githubusercontent.com/51886077/85207417-dc0b1900-b328-11ea-9c71-9251b6c57cca.png)
![image](https://user-images.githubusercontent.com/51886077/85207425-e2999080-b328-11ea-87c5-b112edcd8f59.png)
![image](https://user-images.githubusercontent.com/51886077/85207426-e62d1780-b328-11ea-82d2-7456af572d1c.png)
![image](https://user-images.githubusercontent.com/51886077/85207427-e88f7180-b328-11ea-8e3d-c4388f62aebf.png)
![image](https://user-images.githubusercontent.com/51886077/85207429-eaf1cb80-b328-11ea-8970-7644e8106cf8.png)
![image](https://user-images.githubusercontent.com/51886077/85207431-edecbc00-b328-11ea-9da0-c67affecd3ef.png)
![image](https://user-images.githubusercontent.com/51886077/85207432-f1804300-b328-11ea-8d81-9261185e9a76.png)
![image](https://user-images.githubusercontent.com/51886077/85207433-f34a0680-b328-11ea-9b25-e92f3de1bb80.png)
![image](https://user-images.githubusercontent.com/51886077/85207434-f5ac6080-b328-11ea-9d80-38f6cc42a222.png)
![image](https://user-images.githubusercontent.com/51886077/85207435-f7762400-b328-11ea-9e3a-0d2e3d630cd9.png)
![image](https://user-images.githubusercontent.com/51886077/85207438-f9d87e00-b328-11ea-9cbe-f7a9db1f90d6.png)
![image](https://user-images.githubusercontent.com/51886077/85207439-fba24180-b328-11ea-99cc-b4a19d292572.png)
![image](https://user-images.githubusercontent.com/51886077/85207442-fe049b80-b328-11ea-80f1-ed8cc8b4cab9.png)
![image](https://user-images.githubusercontent.com/51886077/85207444-ffce5f00-b328-11ea-95ba-4280d1620f24.png)

stephaneghozzi commented:
So to summarize the remaining issues, hopefully I'm not overlooking anything else: 
- the new `pred_1` and the data generated for the plots probably should be identical to the original ones (intriguing but maybe not that much of a problem?)
- the tests for the plots fail, presumably for technical reasons regarding saving in svg and using `vdiffr::expect_doppelganger` (the plots themselves look OK)

thibautjombart commented:
Again, a huge thanks for this. I really like the way you show the comparison on the PR, and appreciate that's a lot of time just for this part. But really cool. 

### Comments
* for the plots: I get the same as you; I will scape off the old plots and re-generate the ref svg so hopefully this will go away
* for the objects that are different and should not: weird, it almost looks like a difference in random seed

### What I'll do today
* check and fix the plot references
* code-review the new code
* investigate the remaining issue with ref objects which changed and should not have

I'm hopeful we can merge this soon! :) This will probably win the PR award of the year!

thibautjombart commented:
I am continuing this PR at https://github.com/reconhub/projections/pull/31 which now uses a branch from the project. @stephaneghozzi you now have write access on the repo (sorry for not granting this before, I just realised). I am working on tests etc, so leave it with me for now, more in a few hours. :)
